### PR TITLE
[SPARK-25493][SQL] Use auto-detection for CRLF in CSV datasource multiline mode

### PR DESCRIPTION
**Pull request by justinuang**
## What changes were proposed in this pull request?

CSVs with windows style crlf ('\r\n') don't work in multiline mode. They work fine in single line mode because the line separation is done by Hadoop, which can handle all the different types of line separators. This PR fixes it by enabling Univocity's line separator detection in multiline mode, which will detect '\r\n', '\r', or '\n' automatically as it is done by hadoop in single line mode.

## How was this patch tested?

Unit test with a file with crlf line endings.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 1]
```diff
@@ -212,6 +212,7 @@ class CSVOptions(
     settings.setEmptyValue(emptyValueInRead)
     settings.setMaxCharsPerColumn(maxCharsPerColumn)
     settings.setUnescapedQuoteHandling(UnescapedQuoteHandling.STOP_AT_DELIMITER)
+    settings.setLineSeparatorDetectionEnabled(true)
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: The auto-detection mechanism is enabled for both - multi-line and per-line mode. I guess it has some overhead on detection of new lines which is not needed in per-line mode. I would benchmark it in both modes (see `CSVBenchmarks`), and if the overhead in per-line mode is significant, I would not enable the option when `multiLine` is set to `false`.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: Yup, I would rather enable this only for multiline mode. Also, please add what this configuration does in the PR description.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
Also, please fix the PR title to be more descriptive. For instance, `[SPARK-25493][SQL] Use auto-detection for CRLF in CSV datasource multiline mode`.

**Comment by justinuang** [comment]
It looks like a flake? Can someone retrigger it?
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96511/console

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Mind explaining what `setLineSeparatorDetectionEnabled` does in the PR description as well?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 2]
```diff
@@ -212,6 +212,11 @@ class CSVOptions(
     settings.setEmptyValue(emptyValueInRead)
     settings.setMaxCharsPerColumn(maxCharsPerColumn)
     settings.setUnescapedQuoteHandling(UnescapedQuoteHandling.STOP_AT_DELIMITER)
+
+    if (multiLine) {
+      settings.setLineSeparatorDetectionEnabled(true)
```
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: Would be simple just `settings.setLineSeparatorDetectionEnabled(multiLine)` or `settings.setLineSeparatorDetectionEnabled(multiLine == true)`?

**Comment by justinuang** [comment]
Sounds good, thanks guys =)

**Comment by justinuang** [comment]
What does it take to get this to be merged in?

**Comment by mccheah** [comment]
@HyukjinKwon is this ready to be merged in, or is there more feedback to be addressed?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala [hunk 3]
```diff
@@ -212,6 +212,8 @@ class CSVOptions(
     settings.setEmptyValue(emptyValueInRead)
     settings.setMaxCharsPerColumn(maxCharsPerColumn)
     settings.setUnescapedQuoteHandling(UnescapedQuoteHandling.STOP_AT_DELIMITER)
+    settings.setLineSeparatorDetectionEnabled(multiLine)
```
[the remark this answers is not in the document]
Comment (justinuang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVOptions.scala]: done

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
I haven't checked what `setLineSeparatorDetectionEnabled` does explicitly yet in Univocity parser. Is this exactly same behaviour when we read it via Hadoop's `LineRecordReader`? Also how does it work with `setLineSeparator`? Essentially we should expose this option too.

**Comment by justinuang** [comment]
So Hadoop's LineReader looks like it handles CR, LF, CRLF:

https://github.com/apache/hadoop/blob/f90c64e6242facf38c2baedeeda42e4a8293e642/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/LineReader.java#L36

Univocity handles CR, LF, CRLF (the logic is a bit convoluted but it looks like they have the same behavior in that if they see a CR, they will look for a LF next):

https://github.com/uniVocity/univocity-parsers/blob/master/src/main/java/com/univocity/parsers/common/input/LineSeparatorDetector.java

I do agree we should expose the option of `setLineSeparator`, but regardless of that, the default behavior of handling CR, LF, CRLF should be the same between single line and multiline mode.

**Comment by HyukjinKwon** [comment]
@justinuang, okay. Mind rebasing this please?

**Comment by justinuang** [comment]
done!

File: sql/core/src/test/resources/test-data/cars-crlf.csv [hunk 4]
```diff
@@ -0,0 +1,7 @@
+
+year,make,model,comment,blank
+"2012","Tesla","S","No comment",
+
+1997,Ford,E350,"Go get one now they are going fast",
+2015,Chevy,Volt
+
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/test-data/cars-crlf.csv]: just for clarification, this file has newline variant combinations, right?

**Review by HyukjinKwon — APPROVED** [review]
looks good except one question

File: sql/core/src/test/resources/test-data/cars-crlf.csv (hunk 4 above)
Comment (justinuang) [reply sql/core/src/test/resources/test-data/cars-crlf.csv]: Yea, if you open it with `vi -b sql/core/src/test/resources/test-data/cars-crlf.csv`, you can see the `^M` characters which according to [this](https://stackoverflow.com/questions/3860519/see-line-breaks-and-carriage-returns-in-editor) represents a CLRF.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 5]
```diff
@@ -220,6 +221,17 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
     }
   }
 
+  test("crlf line separators in multiline mode") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: nit: -> `SPARK-25493: crlf line separators in multiline mode` 

when a PR fixes a specific problem, let's add the jira prefix in the test name next time.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
@justinuang, this might affect existing users application. Although this matches the behaviour to non-miltiline mode, can we explicitly mention it in migration guide?

cc @cloud-fan and @gatorsmile

**Comment by jonathanneo** [comment]
@HyukjinKwon CSVs with windows style CR-LF ('\r\n') still doesn't work for me. 

I am using Spark 2.4.3 and using the PySpark API. 

File: [test123-CRLF.zip](https://github.com/apache/spark/files/3349006/test123-CRLF.zip)

When I run the following: 
`dfCSV = sqlContext.read.format("csv").options(header="true", inferSchema="true", delimiter=",", encoding="UTF-8",escape='"', multiLine="true").load("test123-CRLF.csv") 
print(dfCSV.first()) # print the first row`

It returns: 
Row(Test1='hello', Test2 ='world\r')

**Comment by MaxGekk** [comment]
@jonathanneo This was merged to the master, and will be released with Spark 3.0

**Comment by jonathanneo** [comment]
@MaxGekk Thanks Max. Is there a known workaround in the meantime?

**Pull request closed without merging**